File: test_permission_ws.py
import asyncio
import websockets
import json
import sys

async def test_permission_flow():
    uri = "ws://localhost:8001/ws?token=test"

    async with websockets.connect(uri) as ws:
        print("✓ Connected to WebSocket")

        # 1. 创建会话
        create_req = {
            "action": "session_create",
            "title": "Test Permission Flow",
            "cwd": "/Users/wust_lh/MobileVC"
        }
        await ws.send(json.dumps(create_req))
        print(f"→ Sent: {create_req}")

        # 等待会话创建响应
        session_id = None
        while True:
            msg = await ws.recv()
            data = json.loads(msg)
            event_type = data.get('type', 'unknown')
            print(f"← Received: {event_type}")
            if event_type == "session_created":
                print(f"   Full data: {json.dumps(data, indent=2)}")
                session_id = data.get("sessionId") or data.get("id")
                print(f"✓ Session created: {session_id}")
                break

        if not session_id:
            print("✗ Failed to create session")
            return

        # 2. 启动 Claude CLI (使用 exec 而不是 input)
        exec_req = {
            "action": "exec",
            "cmd": "claude",
            "cwd": "/Users/wust_lh/MobileVC",
            "mode": "pty",
            "engine": "claude",
            "command": "claude",
            "permissionMode": "default"
        }
        await ws.send(json.dumps(exec_req))
        print(f"→ Sent exec: {exec_req}")

        # 等待 Claude CLI 启动
        await asyncio.sleep(2)

        # 3. 发送输入触发权限请求（使用 Edit 操作）
        input_req = {
            "action": "input",
            "data": "帮我在 /Users/wust_lh/MobileVC/hhh.txt 文件末尾添加一行：新增内容测试",
            "permissionMode": "default"
        }
        await ws.send(json.dumps(input_req))
        print(f"→ Sent input: {input_req}")

        # 4. 监听权限请求事件
        while True:
            msg = await ws.recv()
            data = json.loads(msg)
            event_type = data.get("type", "unknown")
            print(f"← Received: {event_type}")

            if event_type == "prompt_request":
                print(f"✓ Got permission prompt: {data.get('message', '')}")
                permission_request_id = data.get("runtimeMeta", {}).get("permissionRequestId", "")
                print(f"  Permission Request ID: {permission_request_id}")

                # 5. 批准权限
                decision_req = {
                    "action": "permission_decision",
                    "decision": "approve",
                    "permissionRequestId": permission_request_id
                }
                await ws.send(json.dumps(decision_req))
                print(f"→ Sent decision: {decision_req}")

            elif event_type == "error":
                print(f"✗ Error: {data.get('message', '')}")
                break

            elif event_type == "agent_state":
                state = data.get("state", "")
                message = data.get("message", "")
                print(f"  Agent state: {state} - {message}")
                if state == "idle":
                    print("✓ Agent completed")
                    break

if __name__ == "__main__":
    asyncio.run(test_permission_flow())

File: tests/manual/test_permission_ws.js
#!/usr/bin/env node

const WebSocket = require('ws');

const WS_URL = 'ws://localhost:8001/ws?token=dev';
const SESSION_ID = 'test-session-' + Date.now();
const CWD = '/Users/wust_lh/MobileVC';

console.log('=== 连接 WebSocket ===');
const ws = new WebSocket(WS_URL);

let permissionRequestId = null;

ws.on('open', () => {
  console.log('WebSocket 已连接\n');

  // 1. 加载会话
  console.log('=== 1. 发送 session_load ===');
  ws.send(JSON.stringify({
    type: 'session_load',
    sessionId: SESSION_ID,
    cwd: CWD
  }));

  // 2. 等待 2 秒后发送消息
  setTimeout(() => {
    console.log('\n=== 2. 发送消息（触发权限请求）===');
    ws.send(JSON.stringify({
      type: 'input',
      data: '帮我读取 /Users/wust_lh/MobileVC/hhh.txt 文件',
      permissionMode: 'default'
    }));
  }, 2000);
});

ws.on('message', (data) => {
  try {
    const event = JSON.parse(data.toString());

    // 只打印关键事件
    if (event.type === 'prompt_request') {
      console.log('\n=== 收到 PromptRequestEvent ===');
      console.log('Message:', event.message);
      console.log('Choices:', event.choices);
      console.log('RuntimeMeta:', JSON.stringify(event.runtimeMeta, null, 2));

      permissionRequestId = event.runtimeMeta?.permissionRequestId;

      if (permissionRequestId) {
        // 等待 2 秒后批准权限
        setTimeout(() => {
          console.log('\n=== 3. 发送权限批准 ===');
          console.log('Permission Request ID:', permissionRequestId);
          ws.send(JSON.stringify({
            type: 'permission_decision',
            decision: 'approve',
            permissionRequestId: permissionRequestId,
            permissionMode: 'default'
          }));
        }, 2000);
      }
    } else if (event.type === 'log') {
      console.log('[LOG]', event.text?.substring(0, 100));
    } else if (event.type === 'assistant_message') {
      console.log('[ASSISTANT]', event.text?.substring(0, 100));
    } else if (event.type === 'runtime_phase') {
      console.log('[PHASE]', event.phase, event.kind, event.message);
    } else if (event.type === 'step_update') {
      console.log('[STEP]', event.tool, event.status, event.target);
    } else if (event.type === 'session_end') {
      console.log('\n=== 会话结束 ===');
      console.log('Reason:', event.reason);
      ws.close();
      process.exit(0);
    }
  } catch (e) {
    console.error('解析消息失败:', e.message);
  }
});

ws.on('error', (error) => {
  console.error('WebSocket 错误:', error.message);
  process.exit(1);
});

ws.on('close', () => {
  console.log('\nWebSocket 已关闭');
  process.exit(0);
});

// 30 秒后超时
setTimeout(() => {
  console.log('\n=== 测试超时 ===');
  ws.close();
  process.exit(1);
}, 30000);
